### DISCUSSION TITLE
Usage of Flutter Modular to deal with dependency injection, routing and auto-dispose

### DISCUSSION THREAD

**Discussion by pedrolemoz** [olmps/memo#20] [Ideas (English)]
### Usage of Flutter Modular in the project

In the Brazilian comunnity, Flutter Modular is one of the most used packages. This package show us an approach to modularize our code in a manner that code reutilization is the main focus. Also, the package has it's own abstraction of the original Navigator class, that don't need the actual context to work properly (this is very appropriated when we want to switch between pages using the app controller).

You can see the official [docs here](https://pub.dev/packages/flutter_modular).

The usage is very simple: we have the main.dart file executing the runApp function, as expected. But, instead of passing a Stateless/Stateful MaterialApp widget, we just pass a AppModule widget.

![image](https://user-images.githubusercontent.com/20285104/113033874-e126af80-9167-11eb-86fe-d5b1606dcf3a.png)

In this project, I have two different flavors: production and development. Each flavor has it's own main.dart file, and that's why I'm passing a specific ServerConfiguration in the constructor of AppModule. This is not usual, but works great.

Then, the AppModule will look like this:

![image](https://user-images.githubusercontent.com/20285104/113034413-6e6a0400-9168-11eb-86b7-98d081450ff2.png)

First, we have the binds. This is where we can register our dependencies, and also decide if we want a singleton or not, as well if the bind will be lazy or not.

![image](https://user-images.githubusercontent.com/20285104/113034803-ea644c00-9168-11eb-8daa-668134763326.png)

The i() function is very helpful to get the dependencies in the class constructor, since we don't need to specify the exact instance of the class, or even the type. This process is automatic.

![image](https://user-images.githubusercontent.com/20285104/113035232-665e9400-9169-11eb-8a37-9a07804e53c9.png)

Then, we'll have the ChildModules. These modules are usually sections of the app. For example, we can have an authentication module, home module, cart module, etc... We'll define all the ChildModules in the routers of AppModule.

![image](https://user-images.githubusercontent.com/20285104/113035650-d10fcf80-9169-11eb-946c-b28d76fd080c.png)

Finally, our app entry point will be the bootstrap widget from the AppModule. In this case, AppWidget.

![image](https://user-images.githubusercontent.com/20285104/113035857-fdc3e700-9169-11eb-9d2f-184a547cc517.png)

Our ChildModule will look like this:

![image](https://user-images.githubusercontent.com/20285104/113035952-1207e400-916a-11eb-981c-4ecbd95a12a4.png)

The bindings are similar to those defined in AppModule. The difference here is the scope: all the bindings defined in AppModule will remain until the app is closed. The bindings defined in any child module will only be used inside that module, and will be disposed when the module is closed (usually when the user is not accessing a page within that module). This give us a better overall performance by optimizing the used memory.

In order to access a binding, we use the `Modular.get<T>()` function.

![image](https://user-images.githubusercontent.com/20285104/113036843-213b6180-916b-11eb-9658-15ac51395995.png)

Since we don't need a context to navigate between pages, is safe to use the controller for this:

![image](https://user-images.githubusercontent.com/20285104/113037095-6a8bb100-916b-11eb-9b6b-61b24a676142.png)

We can pass strings into the route:

![image](https://user-images.githubusercontent.com/20285104/113037492-cf470b80-916b-11eb-9a55-e11657cde8f1.png)

![image](https://user-images.githubusercontent.com/20285104/113037574-e84fbc80-916b-11eb-92aa-5cba0f1e755c.png)

And we also can pass objects:

![image](https://user-images.githubusercontent.com/20285104/113038213-a6734600-916c-11eb-874d-959dae0d7961.png)

This was a brief explanation of the package and it's benefits. More details can be found in the internet, and also in the Brazilian community, as the package is very used.

**Comment by matuella** [comment]
Hello @pedrolemoz, and thanks for your contribution!

So, I took the time to read your suggestion and more about the Flutter Modular package, and I found it to be quite an interesting package, so thanks for bringing this up for discussion.

I liked the fact the it is a solution that tackles many problems at the same time, but the relevant ones to this use in this project I think it would be **routing** and **dependency injection**. The problem that I see to include this package and use throughout the project is:
1. we have decided to use `river_pod`, which we are using to provide all dependencies to multiple, and we think it's a lightweight solution that is an improvement over the classic `provider` and `bloc`, you can read more about the package in [its repository](https://github.com/rrousselGit/river_pod);
2. we implemented the navigator using the navigator 2.0 approach. While we didn't quite liked it (and it appears that neither [the Flutter team did](https://github.com/flutter/uxr/wiki/Navigator-2.0-API-Usability-Research), so they are doing a re-work on it), we still think that it's easy to scale to complex flows.

I think is a really good package for a good amount of applications out there, but given the dependencies that we are using to provide DI (that is detached from the Flutter framework) and the _native_ Navigator 2.0 (or Router) solution, we will stick to these until either they prove to be a burden other than a help, or if there is a clear better package (or native solution) to achieve the aforementioned goals.

Again, thanks for taking your time to help improving memo!

**Reply by luizwalber** [reply]
Modular 3.0 is build using the Navigator 2.0

The Navigator 2.0 is complex, so the flutter team is analyzing the packages build based on Navigator, 

so far Modular 3.0 doesn't meet only one of the requirements (Skipping Stacks), you can [check the discussion here](https://github.com/flutter/uxr/tree/master/nav2-usability/comparative-analysis)
